From 4d998cd1e5a9623a9cf2a8ce283d06bcb17537cc Mon Sep 17 00:00:00 2001
From: agaraix <anath.bandux.garai@intel.com>
Date: Wed, 29 Jan 2020 22:22:26 +0530
Subject: [PATCH] UGW_SW-46847 : Software acceleration performance enhance

Avoid raw socket delivery processing for the software accelerated packet.
---
 include/linux/extmark.h | 4 ++++
 net/core/dev.c          | 3 +++
 2 files changed, 7 insertions(+)

diff --git a/include/linux/extmark.h b/include/linux/extmark.h
index 0944bc0306bd..47f054670a40 100644
--- a/include/linux/extmark.h
+++ b/include/linux/extmark.h
@@ -35,6 +35,10 @@
 #define LROSESS_NO_OF_BITS 1
 #define LROSESS_MASK 0x80000000
 
+/* LRO bit is being reused to mark packet is acclerated by SAE */
+#define FILTERTAP_START_BIT_POS 31
+#define FILTERTAP_NO_OF_BITS 1
+#define FILTERTAP_MASK 0x80000000
 
 #define GET_DATA_FROM_MARK_OPT(mark, mask, pos, value) \
 	(value = ((mark & mask) >> pos))
diff --git a/net/core/dev.c b/net/core/dev.c
index cdc965c38e65..bbd0b8229d73 100644
--- a/net/core/dev.c
+++ b/net/core/dev.c
@@ -2973,6 +2973,9 @@ static int xmit_one(struct sk_buff *skb, struct net_device *dev,
 	unsigned int len;
 	int rc;
 
+#ifdef CONFIG_NETWORK_EXTMARK
+	if (!(skb->extmark & FILTERTAP_MASK))
+#endif
 	if (!list_empty(&ptype_all) || !list_empty(&dev->ptype_all))
 		dev_queue_xmit_nit(skb, dev);
 
